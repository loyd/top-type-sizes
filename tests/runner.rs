use top_type_sizes::*;

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            static SAMPLE: &'static str =
                include_str!(concat!("samples/", stringify!($name), ".txt"));

            let refined_sample = reader::read(SAMPLE.as_bytes()).unwrap();
            let types = parser::parse(&refined_sample).unwrap();
            insta::assert_debug_snapshot!(format!("{}_types", stringify!($name)), types);

            // TODO: check the transformer.

            let output = formatter::format(types, &Default::default());
            println!("{output}");
            // TODO: make snapshots of output.
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
