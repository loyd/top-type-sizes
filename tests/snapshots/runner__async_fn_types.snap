---
source: tests/runner.rs
expression: types
---
[
    Type {
        name: "[async fn body@$DIR/async.rs:8:36: 11:2]",
        size: 16386,
        align: 1,
        kind: Enum(
            EnumType {
                discriminant_size: Some(
                    1,
                ),
                variants: [
                    EnumVariant {
                        name: "Unresumed",
                        size: 8192,
                        items: [
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "arg",
                                    size: 8192,
                                    align: Some(
                                        1,
                                    ),
                                    offset: Some(
                                        0,
                                    ),
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Suspend0",
                        size: 16385,
                        items: [
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "arg",
                                    size: 8192,
                                    align: Some(
                                        1,
                                    ),
                                    offset: Some(
                                        0,
                                    ),
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: GeneratorLocal,
                                    name: "arg",
                                    size: 8192,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: GeneratorLocal,
                                    name: "__awaitee",
                                    size: 1,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Returned",
                        size: 8192,
                        items: [
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "arg",
                                    size: 8192,
                                    align: Some(
                                        1,
                                    ),
                                    offset: Some(
                                        0,
                                    ),
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Panicked",
                        size: 8192,
                        items: [
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "arg",
                                    size: 8192,
                                    align: Some(
                                        1,
                                    ),
                                    offset: Some(
                                        0,
                                    ),
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                ],
            },
        ),
        end_padding: None,
    },
    Type {
        name: "std::mem::ManuallyDrop<[u8; 8192]>",
        size: 8192,
        align: 1,
        kind: Struct(
            StructType {
                items: [
                    Field(
                        Field {
                            kind: AdtField,
                            name: "value",
                            size: 8192,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                ],
            },
        ),
        end_padding: None,
    },
    Type {
        name: "std::mem::MaybeUninit<[u8; 8192]>",
        size: 8192,
        align: 1,
        kind: Enum(
            EnumType {
                discriminant_size: None,
                variants: [
                    EnumVariant {
                        name: "MaybeUninit",
                        size: 8192,
                        items: [
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "uninit",
                                    size: 0,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "value",
                                    size: 8192,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                ],
            },
        ),
        end_padding: None,
    },
    Type {
        name: "[async fn body@$DIR/async.rs:6:17: 6:19]",
        size: 1,
        align: 1,
        kind: Enum(
            EnumType {
                discriminant_size: Some(
                    1,
                ),
                variants: [
                    EnumVariant {
                        name: "Unresumed",
                        size: 0,
                        items: [],
                    },
                    EnumVariant {
                        name: "Returned",
                        size: 0,
                        items: [],
                    },
                    EnumVariant {
                        name: "Panicked",
                        size: 0,
                        items: [],
                    },
                ],
            },
        ),
        end_padding: None,
    },
    Type {
        name: "std::mem::ManuallyDrop<[async fn body@$DIR/async.rs:6:17: 6:19]>",
        size: 1,
        align: 1,
        kind: Struct(
            StructType {
                items: [
                    Field(
                        Field {
                            kind: AdtField,
                            name: "value",
                            size: 1,
                            align: None,
                            offset: None,
                            local_type: None,
                        },
                    ),
                ],
            },
        ),
        end_padding: None,
    },
    Type {
        name: "std::mem::MaybeUninit<[async fn body@$DIR/async.rs:6:17: 6:19]>",
        size: 1,
        align: 1,
        kind: Enum(
            EnumType {
                discriminant_size: None,
                variants: [
                    EnumVariant {
                        name: "MaybeUninit",
                        size: 1,
                        items: [
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "uninit",
                                    size: 0,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "value",
                                    size: 1,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                ],
            },
        ),
        end_padding: None,
    },
    Type {
        name: "std::task::Poll<()>",
        size: 1,
        align: 1,
        kind: Enum(
            EnumType {
                discriminant_size: Some(
                    1,
                ),
                variants: [
                    EnumVariant {
                        name: "Ready",
                        size: 0,
                        items: [
                            Field(
                                Field {
                                    kind: AdtField,
                                    name: "0",
                                    size: 0,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Pending",
                        size: 0,
                        items: [],
                    },
                ],
            },
        ),
        end_padding: None,
    },
]
